packages/jupyter-ai/jupyter_ai/actors/autonotebook.py

```diff
@@ -0,0 +1,235 @@
+            "Content:\n{content}"
+        )
+        prompt = PromptTemplate(
+            template=task_creation_template,
+            input_variables=[
+                "content",
+            ],
+        )
+        return cls(prompt=prompt, llm=llm, verbose=verbose)
+
+def generate_title_and_summary(outline, llm=None, verbose=False):
+    if llm is None:
+        llm = ChatOpenAINewProvider(model_id='gpt-3.5-turbo')
+    summary_chain = NotebookSummaryChain.from_llm(llm=llm, verbose=verbose)
+    title_chain = NotebookTitleChain.from_llm(llm=llm, verbose=verbose)
+    summary = summary_chain.predict(content=outline)
+    title = title_chain.predict(content=outline)
+    outline['summary'] = summary
+    outline['title'] = title.strip('"')
+    return outline
+
+def create_notebook(outline):
+    nb = nbformat.v4.new_notebook()
+    markdown_cell = nbformat.v4.new_markdown_cell('# ' + outline['title'])
+    nb['cells'].append(markdown_cell)
+    markdown_cell = nbformat.v4.new_markdown_cell('## Introduction')
+    nb['cells'].append(markdown_cell)
+    disclaimer = f"This notebook was created by [Jupyter AI](https://github.com/jupyterlab/jupyter-ai) with the following prompt:\n\n> {outline['description']}"
+    markdown_cell = nbformat.v4.new_markdown_cell(disclaimer)
+    nb['cells'].append(markdown_cell)
+    markdown_cell = nbformat.v4.new_markdown_cell(outline['summary'])
+    nb['cells'].append(markdown_cell)
+
+    for section in outline['sections'][1:]:
+        markdown_cell = nbformat.v4.new_markdown_cell('## ' + section['title'])
+        nb['cells'].append(markdown_cell)
+        for code_block in section['code'].split('\n\n'):
+            code_cell = nbformat.v4.new_code_cell(code_block)
+            nb['cells'].append(code_cell)
+        nb['cells'].append(code_cell)
+    return nb
+
+@ray.remote
+class AutoNotebookActor(BaseActor):
+    def __init__(self, reply_queue: Queue, root_dir: str, log: Logger):
+        super().__init__(log=log, reply_queue=reply_queue)
+        self.root_dir = os.path.abspath(os.path.expanduser(root_dir))
+        self.llm = ChatOpenAINewProvider(model_id='gpt-3.5-turbo')
+  
+    def process_message(self, message: HumanChatMessage):
+        agent_message = AgentChatMessage(
+                id=uuid4().hex,
+                time=time.time(),
+                body="👍 Great, I will get started on your notebook. It may take a few minutes, but I will reply here when the notebook is ready. In the meantime, you can continue to ask me other questions.",
+                reply_to=message.id
+        )
+        self.reply_queue.put(agent_message)
+        outline = generate_outline(message.body, llm=self.llm, verbose=True)
+        outline = generate_code(outline, llm=self.llm, verbose=True)
+        outline = generate_title_and_summary(outline, llm=self.llm)
+        notebook = create_notebook(outline)
+        final_path = os.path.join(self.root_dir, outline['title'] + '.ipynb')
```

Comment:
Should we have a separate variable for the filename, so that we can strip out unsuitable characters like `/` or `&` (on Windows), and so that we can replace spaces with `-` or `_`?

Reply:
Yep, let's do this in a follow up PR.